File: docs/sentence-similarity/sample.java
import org.apache.commons.text.similarity.CosineSimilarity;
import java.util.*;

public class QuoteSearcher {
    public static void main(String[] args) {
        // Corpus of quotations
        List<String> corpus = Arrays.asList(
            "The only limit to our realization of tomorrow is our doubts of today.",
            "In the middle of every difficulty lies opportunity.",
            "Success is not final, failure is not fatal: It is the courage to continue that counts."
        );

        // Input query
        String query = "Every difficulty brings opportunity.";

        // Preprocess corpus and query (example: lowercase and tokenization)
        Map<CharSequence, Integer> queryVector = createTFVector(query);
        List<String> results = new ArrayList<>();

        for (String quote : corpus) {
            Map<CharSequence, Integer> quoteVector = createTFVector(quote);
            double similarity = calculateCosineSimilarity(queryVector, quoteVector);

            // Filter results by a similarity threshold
            if (similarity > 0.2) { // Adjust the threshold as needed
                results.add(quote + " (Similarity: " + similarity + ")");
            }
        }

        System.out.println("Similar Quotes:");
        results.forEach(System.out::println);
    }

    private static Map<CharSequence, Integer> createTFVector(String text) {
        Map<CharSequence, Integer> vector = new HashMap<>();
        String[] words = text.toLowerCase().split("\\W+");
        for (String word : words) {
            vector.put(word, vector.getOrDefault(word, 0) + 1);
        }
        return vector;
    }

    private static double calculateCosineSimilarity(Map<CharSequence, Integer> vec1, Map<CharSequence, Integer> vec2) {
        CosineSimilarity cosine = new CosineSimilarity();
        return cosine.cosineSimilarity(vec1, vec2);
    }
}

File: docs/sentence-similarity/sample.kt
import ai.djl.Model
import ai.djl.ModelZoo
import ai.djl.modality.nlp.DefaultVocabulary
import ai.djl.modality.nlp.embedding.WordEmbedding
import ai.djl.translate.TranslateException

fun main() {
    val corpus = listOf(
        "The only limit to our realization of tomorrow is our doubts of today.",
        "In the middle of every difficulty lies opportunity.",
        "Success is not final, failure is not fatal: It is the courage to continue that counts."
    )

    val query = "Every difficulty brings opportunity."

    // Load pre-trained embeddings model (example: Universal Sentence Encoder or similar)
    val model = ModelZoo.loadModel("sentence_transformers/all-MiniLM-L6-v2")
    val embedder = model.newPredictor<String, FloatArray>()

    val queryEmbedding = embedder.predict(query)
    val similarities = mutableListOf<Pair<String, Float>>()

    for (quote in corpus) {
        val quoteEmbedding = embedder.predict(quote)
        val similarity = cosineSimilarity(queryEmbedding, quoteEmbedding)
        if (similarity > 0.2) {
            similarities.add(quote to similarity)
        }
    }

    println("Similar Quotes:")
    similarities.sortedByDescending { it.second }.forEach { println("${it.first} (Similarity: ${it.second})") }
}

fun cosineSimilarity(vec1: FloatArray, vec2: FloatArray): Float {
    val dotProduct = vec1.zip(vec2).sumOf { it.first * it.second }
    val magnitude = Math.sqrt(vec1.sumOf { it * it } * vec2.sumOf { it * it })
    return (dotProduct / magnitude).toFloat()
}
